src/patches.js

```diff
@@ -6,50 +6,92 @@ export function generatePatches(state, basePath, patches, inversePatches) {
         : generateObjectPatches(state, basePath, patches, inversePatches)
 }
 
-function generateArrayPatches(state, basePath, patches, inversePatches) {
-    const {base, copy, assigned} = state
-    const minLength = Math.min(base.length, copy.length)
+function _generateArrayPatches(
+    base,
+    copy,
+    assigned,
+    basePath,
+    patches,
+    inversePatches
+) {
+    if (copy.length < base.length) {
+        return _generateArrayPatches(
+            copy,
+            base,
+            assigned,
+            basePath,
+            inversePatches,
+            patches
+        )
+    }
 
-    // Look for replaced indices.
-    for (let i = 0; i < minLength; i++) {
-        if (assigned[i] && base[i] !== copy[i]) {
-            const path = basePath.concat(i)
-            patches.push({op: "replace", path, value: copy[i]})
-            inversePatches.push({op: "replace", path, value: base[i]})
-        }
+    const delta = copy.length - base.length
+
+    let start = 0
+    while (base[start] === copy[start] && start < base.length) {
+        ++start
+    }
+
+    let baseEnd = base.length
+    while (baseEnd > start && base[baseEnd - 1] === copy[baseEnd + delta - 1]) {
+        --baseEnd
     }
 
-    // Did the array expand?
-    if (minLength < copy.length) {
-        for (let i = minLength; i < copy.length; i++) {
+    const replaceCount = baseEnd - start
+    for (let i = 0; i < replaceCount; ++i) {
+        const index = start + i
+        const path = basePath.concat([index])
+        if (assigned[index] && copy[index] !== base[index]) {
             patches.push({
-                op: "add",
-                path: basePath.concat(i),
-                value: copy[i]
+                op: "replace",
+                path,
+                value: copy[index]
+            })
+            inversePatches.push({
+                op: "replace",
+                path,
+                value: base[index]
             })
         }
-        inversePatches.push({
-            op: "replace",
-            path: basePath.concat("length"),
-            value: base.length
-        })
     }
+    start += replaceCount
 
-    // ...or did it shrink?
-    else if (minLength < base.length) {
+    const useRemove = start != base.length
+    for (let j = 0; j < delta; ++j) {
+        const path = basePath.concat([start + j])
         patches.push({
-            op: "replace",
-            path: basePath.concat("length"),
-            value: copy.length
+            op: "add",
+            path,
+            value: copy[start + j]
         })
-        for (let i = minLength; i < base.length; i++) {
-            inversePatches.push({
-                op: "add",
-                path: basePath.concat(i),
-                value: base[i]
+        if (useRemove) {
+            inversePatches.unshift({
+                op: "remove",
+                path
             })
         }
     }
+
+    if (!useRemove) {
+        inversePatches.push({
```

Comment:
Every inverse patch needs an associated patch, but this doesn't seem to have one. Why?

Also, an inverse patch that replaces the `length` is useless. What about the values that were truncated?

Reply:
It'a alternative for multiple 'remove' operation. So we use only one of them

Set length for arrays is standard practice for JS. Truncated items will be cleared
```
var a = [1,2,3];
a.length = 1;
a.length = 2;
console.log(a); // [1, _empty_]
```

Reply:
This behavior was implemented by @mweststrate, I just kept it
```
describe("arrays - push multiple", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.push(4, 5)
        },
        [
            {op: "add", path: ["x", 3], value: 4},
            {op: "add", path: ["x", 4], value: 5}
        ],
        [{op: "replace", path: ["x", "length"], value: 3}]
    )
})
```

Reply:
Ah, I missed that test! 👍